### ISSUE TITLE
Incorrect parsing of (NULL) row literal in PostgreSQL for nested records, UDTs, etc.

### ISSUE THREAD

**Issue by atollk** [jOOQ/jOOQ#13169]
(I'm using Kotlin, so my code examples below are also Kotlin, not Java)

**Brief summary**: Postgres returns `null` for a field that contains the string `"null"` but jOOQ parses  this as the SQL `null` literal. Therefore, records can include false `null` values, leading to NullPointerExceptions in code.

### Expected behavior

Assume that, in our Postgres db, we have a table `X` with a column `c` of type `text` (`not null`, although I don't believe this matters). For simplicity, I'll also assume that `c` is the only column of the table for this example.

I expect the following query to return a single instance of `Record1<Record1<String>>`:
`val result = select(table.fieldRow()).from(table).fetchOne()!!`
such that `result.value1().value2()` will be the String value of the column `c` of some row in the table `X`.

### Actual behavior

Continuing the example above, if the row selected by the query has the text value `null` (i.e., not SQL null but the text "null" without quotation marks), the actual value of `result.value1().value1()` will be `null` (Java/Kotlin null). In fact, this happens for any text value that starts with the string "null", independent of casing.

### Steps to reproduce the problem

```sql
create table messages
(
    id int not null primary key,
    body text not null,
);

insert into messages values (123, "null");
```

```kotlin
val record = DSL.select(MESSAGES.fieldsRow()).from(MESSAGES).fetchOne()
val body = record!!.value1().value2()
body!!  // NullPointerException
```

Note that this only happens when we use `select(table.fieldsRow())`, not with `selectFrom(table)`, `selct().from(table)`, or `select(table.asterisk())`.

If this explanation is not sufficient, let me know and I'll create a MCVE.

### Versions

- jOOQ: 3.16.4 Open Source
- Java: 14
- Kotlin: 1.5.21
- Database: Postgres 42.2.16
- OS: Arch Linux

### Additional information

A query such as `select (messages.body, null, "null", "")` in psql shows how Postgres treats the different values:

```
       row       
-----------------
 (null,,null,"")
```

An actual SQL null is shown as nothing, whereas the empty string has two quotation marks, and the string "null" is shown as `null` without quotation marks. 
I believe that this piece of code is what's parsing the result incorrectly: https://github.com/jOOQ/jOOQ/blob/main/jOOQ/src/main/java/org/jooq/util/postgres/PostgresUtils.java#L451

This use case has become relevant to us with the arrival of Multiset and adhoc conversion in jOOQ 3.15, as we'd like to load one-to-many relations in general like so:

```kotlin
val filteredTable = MY_TABLE.where(/* ... */).orderBy(/* ... */)
val filteredTableWithRelations = select(
    filteredTable.fieldsRow().`as`(MY_TABLE.name()).convertFrom { /* ... */ },
    multiset(/* ... */)
)
// filteredTableWithRelations now contains MY_TABLE-records as well as their required relations from other tables.
```

**Comment by lukaseder** [comment]
Thanks a lot for your detailed report. I don't think any additional MCVE is needed, I can see how this might go wrong. Will look into this tomorrow.

**Comment by atollk** [comment]
Cheers, I appreciate it. To me, it'd be interesting to know in which release (3.16.5, 3.16.6, 3.17.0, or something else) you realistically see a fix possible, so I know how much effort I should spend on finding a workaround :)

**Comment by lukaseder** [comment]
> To me, it'd be interesting to know in which release (3.16.5, 3.16.6, 3.17.0, or something else) you realistically see a fix possible

Ah yes, the good old software time and scope estimate request. I don't even *know* if the fix is trivial or complex, so let me express the only logical answer: ASAP? 😅

> I know how much effort I should spend on finding a workaround

I might help you with that? The workaround for your *specific* example is obvious: Don't nest the row. But what might be your *actual* query?

**Comment by lukaseder** [comment]
At least, it was easy to reproduce. Investigating now.

**Comment by lukaseder** [comment]
I was going to check if this problem happens to UDTs as well, which use the same string representation, and for which the linked `PostgresUtils` logic was implemented, but then I found another bug: https://github.com/jOOQ/jOOQ/issues/13174

**Comment by lukaseder** [comment]
So, I agree, that logic in `PostgresUtils`, which consumes the `"null"` string, seems faulty. `null` values are encoded in that PG object notation as an absent value, e.g. `()`, (,)`, etc.

However, there are cases where (regrettably) this notation seems inconsistent, e.g. when arrays are involved:

```sql
select array[row(null, null), null]
```

Produces

```
{"(,)",NULL}
```

Though:

```sql
select array['a', null]
```

Produces

```
{a,}
```

I guess that's where the current logic comes from. Removing the logic creates a failure in other integration tests, so we need to tread carefully here.

**Comment by lukaseder** [comment]
Hmm, it seems that `NULL` always means the `NULL` value in arrays, whereas `"NULL"` is the `'NULL'` string:

```
postgres=# select
  array[''],
  array[]::text[],
  row('', 'NULL'),
  row(null, 'NULL'),
  array[null, 'NULL'],
  row(array[null]),
  row(array[null, 'NULL']),
  array[row(null), row('NULL')];

 array | array |    row    |   row   |     array     |   row    |         row         |    array
-------+-------+-----------+---------+---------------+----------+---------------------+-------------
 {""}  | {}    | ("",NULL) | (,NULL) | {NULL,"NULL"} | ({NULL}) | ("{NULL,""NULL""}") | {(),(NULL)}
```

Maybe, psql is more of an authority here than DBeaver.

**Comment by lukaseder** [comment]
Yeah, seems that implementing that `"NULL"` parsing logic in `PostgresUtils` only for array literals, not for row literals does the trick.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.17.0, 3.16.5 (https://github.com/jOOQ/jOOQ/issues/13177), and 3.15.9 (https://github.com/jOOQ/jOOQ/issues/13178)

**Comment by atollk** [comment]
Well, that was fast :) Thank you.
I was just writing down my use case in more detail when you posted the fix. But, unless you want to release the patch versions this week already, I'll still need to find a workaround until then.

> I might help you with that? The workaround for your specific example is obvious: Don't nest the row. But what might be your actual query?

Sure, I appreciate it. The actual use case is what I tried to explain in the last paragraph of the original report. But let my try again:

We have two tables, `messages` and `attachments`. The former has a one-to-many relation to the latter; i.e., a message can contain any number of attachments. 
I'm working in the context of `messages`, i.e. I'm running some queries to fetch data from that table. However, I would like those queries to return not only the `messages` row but also all related rows from `attachments`. This can be imagined as something like a `Pair<MessagesRecord, List<AttachmentsRecord>>`.
For ease of use, the goal is to offer a function like `loadAttachments`, which "enriches" the data from a `messages` row by the related data from `attachments`. A usage example would be as such:

```kotlin
val queryMessagesWithEmptyBody: Select<MessagesRecord> =
    dsl.selectFrom(MESSAGES).where(MESSAGES.body.eq(""))
val messageWithAttachment: Pair<MessagesRecord, List<AttachmentsRecord>> = 
    loadAttachments(queryMessagesWithEmptyBody).fetchOne()
```

The implementation so far for `loadAttachments` was:

```kotlin
fun Select<*>.loadAttachments(): Select<Record2<MessagesRecord, List<AttachmentsRecord>> {
    val messageField = (this.fieldsRow() as SelectField<*>)
	    .`as`(MESSAGES.name)
	    .convertFrom { (it as Record).into(MESSAGES) }
    val attachments = selectAttachmentsViaMultiset(
            ATTACHMENTS.MESSAGE_ID.eq(this.field(MESSAGES.ID))
        )
    return dsl.select(messageField, attachments).from(this)
}
```

I could do something like `messageField = this.asTable().asterisk()` instead but then I'd have to manually extract every single field from the resulting record instead of being able to work with one concise `MessagesRecord`. Also, I lose type safety during compile time.

The context I use the function in also requires that it works on `Select<*>` or `TableLike<*>` (instead of something like a `Table<MessagesRecord>`) and that it returns a `Select` instead of a `Result`.

So, instead, the closest to a working workaround I got so far was to wrap the row in a JSON/JSONB object, like jOOQ does for multisets, but I haven't got the JSON-to-Record parsing running yet.

> However, there are cases where (regrettably) this notation seems inconsistent

That's what I expected was the cause. Great to see that you still managed to fix it as smoothly.
